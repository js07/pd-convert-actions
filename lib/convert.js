const transform = require("./transform");
const generate = require("./generate");
const fix = require("./fix");
const parse = require("./parse");

const PLACEHOLDER_APP_SLUG = "app_placeholder";
const DEFAULT_VERSION_PATCH = 1;

function dbProp() {
  return { key: "db", type: "$.interface.db", isString: true };
}

function paramTypeToPropType(type) {
  const mapping = {
    "array": "any",
    "boolean": "boolean",
    "integer": "integer",
    "number": "integer",
    "object": "object",
    "string": "string",
  };
  if (!mapping[type]) {
    throw new Error(`Can't convert unsupported param type "${type}" to a prop type`);
  }
  return mapping[type];
}

function paramToProp(param, paramKey, required, createLabel) {
  const label = createLabel
    ? (param._label ?? `${key.charAt(0).toUpperCase()}${key.slice(1)}`.replace(/_/g, " "))
    : undefined;
  // If a prop key contains a colon, its object key must be a string literal
  // rather than identifier (e.g. `{ prop:key: {} }` -> `{ "prop:key": {} }`)
  const key = paramKey.includes(":") ? `"${paramKey}"` : paramKey;
  return {
    key,
    type: paramTypeToPropType(param.type),
    label,
    description: param.description,
    optional: !required || undefined
  };
}

function paramsSchemaToProps(params, { createLabel }={}) {
  return Object.entries(params?.properties ?? {}).map(([key, value]) => ({
    ...paramToProp(value, key, params.required.includes(key), createLabel),
  }));
}

function authsToAppProps(auths) {
  return auths.map((appName) => ({
    key: appName,
    type: "app",
    app: appName,
  }));
}

function makeComponentKey(appSlug, componentSlug) {
  return `${appSlug}-${componentSlug}`;
}

function wrapCodeWithFunctionExpr(source) {
  if (source.startsWith("async (")) {
    return source;
  }
  return `async (params, auths) => {\n${source}\n}`;
}

// Wrap code with function Expression
// Get app props from `auths.appName` in code
// If code has this.$checkpoint, add db prop
// Convert params to array of props
// Convert code to component code
// Generate component file using handlebars template, props, and component code
/**
 * 
 * @param {object} actionConfig 
 * @param {object} options 
 * @returns 
 */
async function convert({
  code: codeRaw,
  title,
  description,
  namespace,
  codeConfig: codeConfigString,
  versionMajor = 0,
  versionMinor = 0,
  versionPatch = DEFAULT_VERSION_PATCH,
  hashId,
}, { defineComponent=false, createLabel=false }={}) {
  // const escapedCodeConfigString = codeConfigString.replace(/\\/g, "\\\\");
  const { params_schema: paramsSchema } = JSON.parse(codeConfigString);

  let source = wrapCodeWithFunctionExpr(codeRaw);

  const {
    auths,
    checkpoints
  } = parse(source);

  // Get props from code (`auths.app`) & params
  const appProps = authsToAppProps(auths);
  const appSlug = appProps?.[0]?.key ?? PLACEHOLDER_APP_SLUG;
  const paramsProps = paramsSchemaToProps(paramsSchema, { createLabel });
  
  const props = [
    ...appProps,
    ...(checkpoints.length > 0 ? [dbProp()] : []),
    ...paramsProps,
  ];

  const transformedCode = transform(source);
  
  const componentKey = makeComponentKey(appSlug, namespace);
  const componentCode = generate({
    code: transformedCode,
    props,
    name: title,
    description,
    key: componentKey,
    version: `${versionMajor}.${versionMinor}.${versionPatch}`,
    hashId,
  }, { defineComponent });

  const lintedCode = await fix(componentCode);

  return {
    code: lintedCode,
    appSlug,
    componentSlug: namespace,
  };
}

module.exports = convert;
