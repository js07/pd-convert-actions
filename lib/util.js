import { customAlphabet } from "nanoid";
import fs from "fs";
import path from "path";
import url from "url";
const __dirname = url.fileURLToPath(new URL(".", import.meta.url));

export function fixByteCharacters(string) {
  return string.replace(/[]/g, "'"); // invisible character, supposed to be (’)
}

export function makeFilePath(filePath, relative=false) {
  return relative ? path.join(__dirname, filePath) : filePath;
}

export function readFile(filePath, { relative = false }={}) {
  filePath = makeFilePath(filePath, relative);
  const fileContent = fs.readFileSync(filePath, "utf-8");
  return fileContent;
}

export function writeFile(filePath, content, { relative = false }={}) {
  filePath = makeFilePath(filePath, relative);
  const dirPath = filePath.split("/").slice(0, -1).join("/");
  fs.mkdirSync(dirPath, { recursive: true });
  fs.writeFileSync(filePath, content, "utf-8");
}

export function uniqueId() {
  const nanoid = customAlphabet("1234567890abcdef", 6);
  return nanoid();
}

export default {
  readFile,
  writeFile,
  fixByteCharacters,
  uniqueId,
};