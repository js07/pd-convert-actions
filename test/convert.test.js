import { csvFileToJson } from "../lib/util.js";
import legacyCodeStep from "./data/legacy-code-step";
import convert from "../lib/convert.js";

const parseActionsCsv = async (path) => {
  const actionConfigs = await csvFileToJson(path);
  return actionConfigs.map(actionConfig => ({
    code: actionConfig.CODE_RAW,
    codeConfig: actionConfig.CODE_CONFIG_JSON,
    namespace: actionConfig.DEFAULT_NAMESPACE,
    title: actionConfig.TITLE,
    description: actionConfig.DESCRIPTION,
    versionMajor: actionConfig.PUBLISHED_VERSION_MAJOR,
    versionMinor: actionConfig.PUBLISHED_VERSION_MINOR,
    hashId: actionConfig.HID,
  }));
};

test("Converts published legacy action to component action", async () => {
  const actionConfigs = await parseActionsCsv("./test/data/actions.csv");
  const actionConfig = actionConfigs[0];
  const convertedAction = await convert(actionConfig, {
    defineComponent: false,
    createLabel: false,
    toEsm: true,
    addPlaceholderAppProp: false,
  });
  expect(convertedAction.code).toMatchSnapshot();
});

test("Converts legacy code step to component", async () => {
  const convertedAction = await convert(legacyCodeStep, {
    defineComponent: false,
    createLabel: false,
    toEsm: true,
    usePipedreamLintRules: false,
  });
  expect(convertedAction.code).toMatchSnapshot();
});
