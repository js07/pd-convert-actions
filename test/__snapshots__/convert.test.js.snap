// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Converts legacy code step to component 1`] = `
"import { axios } from "@pipedream/platform";

export default {
  props: {
    trello: {
      type: "app",
      app: "trello",
    },
    db: "data_store",
    myRequiredField: {
      type: "string",
      description: "A required field",
    },
    myOptionalField: {
      type: "string",
      description: "An optional field",
      optional: true,
    },
  },
  async run({ steps, $ }) {
    let someexport;
    let $checkpoint = await this.db.get("$checkpoint");
    let stepCheckpoint = await this.db.get("scp_abc123");
    
    try {
    
      // conversion should use data_store prop to get and set the export at the top
      // and bottom of the run method
      someexport = [];
      someexport.push("asdf");
    
      // $attachments -> step.trigger.context.attachments
      fs.readFileSync(steps.trigger.context.attachments["test4.csv"]).toString();
    
      // should become \`const eventData = steps.trigger.event.data\`
      const eventData = steps.trigger.event.data;
      console.log(eventData);
    
      // conversion should use data_store prop to get and set the checkpoint, with a
      // (mock) unique id for the step, at the top and bottom of the run method
      stepCheckpoint ||= {};
      stepCheckpoint["foo"] = 5;
    
      // conversion should use data_store prop to get and set the checkpoint at the
      // top and bottom of the run method
      $checkpoint ||= {};
      $checkpoint["foo"] = 6;
    
      // conversion should have a "myRequiredField" prop
      console.log(this.myRequiredField);
      // conversion should have a "myOptionalField" prop with optional: true
      console.log(this.myOptionalField);
    
      // should become \`return await axios($, {\` when converting to ESM
      return await axios($, {
        url: \`https://api.trello.com/1/cards/\`+ steps.nodejs.$return_value,
      }, {
        token: {
        // should convert \`auths.trello\` to \`this.trello.$auth\`
          key: this.trello.$auth.oauth_access_token,
          secret: this.trello.$auth.oauth_refresh_token,
        },
        oauthSignerUri: this.trello.$auth.oauth_signer_uri,
      });
    
    } finally {
      await this.db.set("scp_abc123", stepCheckpoint);
      await this.db.set("$checkpoint", $checkpoint);
      $.export("someexport", someexport);
    }
  }
};"
`;

exports[`Converts published legacy action to component action 1`] = `
"// legacy_hash_id: a_2wim5R
import { axios } from "@pipedream/platform";

export default {
  key: "segment-track",
  name: "Track actions your users perform",
  description: "Track lets you record the actions your users perform (note requires userId or anonymousId)",
  version: "0.3.1",
  type: "action",
  props: {
    segment: {
      type: "app",
      app: "segment",
    },
    anonymousId: {
      type: "string",
      description: "A pseudo-unique substitute for a User ID, for cases when you don't have an absolutely unique identifier. A userId or an anonymousId is required.",
      optional: true,
    },
    context: {
      type: "object",
      description: "Dictionary of extra information that provides useful context about a message, but is not directly related to the API call like ip address or locale",
      optional: true,
    },
    event: {
      type: "string",
      description: "Name of the action that a user has performed.",
    },
    integrations: {
      type: "object",
      description: "Dictionary of destinations to either enable or disable",
      optional: true,
    },
    properties: {
      type: "object",
      description: "Free-form dictionary of properties of the event, like revenue",
      optional: true,
    },
    timestamp: {
      type: "string",
      description: "ISO-8601 date string",
      optional: true,
    },
    userId: {
      type: "string",
      description: "Unique identifier for the user in your database. A userId or an anonymousId is required.",
      optional: true,
    },
  },
  async run({ $ }) {
    return await axios($, {
      method: "post",
      url: "https://api.segment.io/v1/track",
      auth: {
        username: this.segment.$auth.write_key,
      },
      data: {
        anonymousId: this.anonymousId,
        context: this.context,
        event: this.event,
        integrations: this.integrations,
        properties: this.properties,
        timestamp: this.timestamp,
        userId: this.userId,
      },
    });
  },
};
"
`;
